### PR TITLE
Detect binary modules even if .node file is missing. Fixes gh-542

### PR DESCRIPTION
**Pull request by rwaldron**
~~I still need to write new tests for this~~

Tests included

### CODE REVIEW

File: lib/tessel/deploy.js [hunk 1]
```diff
@@ -421,20 +440,85 @@ actions.resolveBinaryModules = function(opts) {
...
+        // files can be, take a look at this:
+        //
+        // https://github.com/nodejs/node-gyp/wiki/%22binding.gyp%22-files-out-in-the-wild
+        //
+        var bindingGypPath = path.join(globRoot, modulePath, binName);
+        var bindingGypData;
+        var bindingGypJson;
+
+        try {
+          // Sometimes people write really tidy binding.gyp files that are
```
Comment (johnnyman727) [inline lib/tessel/deploy.js]: This comment makes me simultaneously really happy and really sad.

File: lib/tessel/deploy.js [hunk 2]
```diff
@@ -528,6 +604,21 @@ actions.resolveBinaryModules = function(opts) {
   });
 };
 
+actions.resolveBinaryModules.readGypFileSync = function(gypfile) {
+  var python = process.env.PYTHON || 'python';
+  var program = `import ast, json; print json.dumps(ast.literal_eval(open("${gypfile}").read()));`;
```
Comment (johnnyman727) [inline lib/tessel/deploy.js]: So we've got Python code inside a JavaScript library to pull in a binary likely written in C++/C. Nice.

File: lib/tessel/deploy.js [hunk 3]
```diff
@@ -528,6 +604,21 @@ actions.resolveBinaryModules = function(opts) {
...
+  var decoder = new StringDecoder('utf8');
+  var result = cp.spawnSync(python, ['-c', program]);
+  var output = result.output;
+
+  return output.reduce((accum, buffer) => {
+    if (buffer) {
+      accum += decoder.write(buffer);
+    }
+    return accum;
```
Comment (johnnyman727) [inline lib/tessel/deploy.js]: This function is actually way cleaner than I thought it would end up being.

File: lib/tessel/deploy.js [hunk 4]
```diff
@@ -421,20 +440,85 @@ actions.resolveBinaryModules = function(opts) {
...
+
+            logMissingBinaryModuleWarning(packageJson.name);
+            return bins;
+          }
+        }
+
+        if (bindingGypJson && Array.isArray(bindingGypJson.targets)) {
+          // Anything that can't be covered by this will have to be
+          // dealt with as we encounter them and as they are reported.
+          binName = bindingGypJson.targets[0].target_name + '.node';
```
Comment (johnnyman727) [inline lib/tessel/deploy.js]: Seems reasonable. I don't really understand why there can be multiple targets but that's not really our problem.

**Comment by johnnyman727** [comment]
+100 points for creativity.

File: lib/tessel/deploy.js (hunk 1 above)
Comment (rwaldron) [inline lib/tessel/deploy.js]: Same here my friend

File: lib/tessel/deploy.js (hunk 2 above)
Comment (rwaldron) [inline lib/tessel/deploy.js]: Yep!

:D

File: lib/tessel/deploy.js (hunk 3 above)
Comment (rwaldron) [inline lib/tessel/deploy.js]: Two things really stand out for me: 

- The addition of spawnSync
- Template strings really make a difference

File: lib/tessel/deploy.js (hunk 4 above)
Comment (rwaldron) [inline lib/tessel/deploy.js]: I have no idea either, but I had to draw a line somewhere and this seemed reasonable

**Comment by rwaldron** [comment]
FWIW, I actually made use of my Windows 10 machine to work out the issues in the tests. That was interesting...

**Comment by rwaldron** [comment]
@johnnyman727 @HipsterBrown good to go here?

**Comment by johnnyman727** [comment]
Code looks good to me but I didn't get a chance to test it in action. Do you have a smoke test you use?

**Comment by rwaldron** [comment]
Smoke testing this is pretty convoluted. As I began writing a smoke test for this, I discovered that nested binaries were still getting dropped—this has been fixed.

**Comment by rwaldron** [comment]
## Smoke Test

With a fully booted Tessel 2 running, paste the following into the terminal: 

```sh
mkdir missing-binary-smoke-test && cd missing-binary-smoke-test && t2 init;
npm install bufferutil;
rm node_modules/bufferutil/build/Release/bufferutil.node;
rm node_modules/bufferutil/fallback.js;
echo "var bu = require('bufferutil').BufferUtil; console.log(typeof bu);" > index.js;
t2 run index.js;
```


1. Creates a directory, enters the directory and initializes a tessel project
2. Installs a module the contains a binary
3. Delete the the binary file
4. Delete the "fallback", this ensures that the running script doesn't produce a false positive
5. Populate the index.js file with a simple program
6. Deploy to the tessel


A complete and successful run will _end_ with the following: 

```sh
INFO Looking for your Tessel...
INFO Connected to bishop.
INFO Building project.
INFO Writing project to RAM on bishop (38.912 kB)...
INFO Deployed.
INFO Running index.js...
function
```

**Comment by rwaldron** [comment]
Note that there remains a known issue where modules that contain a binary which is directly "exported" via package.json "main" property are still problematic. T2's deployment has no problem finding the binary and deploying, it but the node process running on the T2 cannot find it; this is out of scope for this project to attempt to fix.

**Comment by johnnyman727** [comment]
:+1: Your smoke tests worked for me. I also did one additional test just for fun which also passed:

```
npm install utf-8-validate
rm node_modules/utf-8-validate/build/Release/validation.node
rm node_modules/utf-8-validate/fallback.js
echo "var u8v = require('utf-8-validate').Validation.isValidUTF8; console.log(typeof u8v);" > index.js
t2 run index.js
```
Output:
```
function
```

Feel free to merge!

**Comment by rwaldron** [comment]
Thanks @johnnyman727 !

**Comment by johnnyman727** [comment]
:pray:

**Pull request merged**